xproc/src/main/xml/specification.xml

```diff
@@ -1921,286 +1921,578 @@ interpreted as follows:</para>
+
+<para>The function may be implemented as an operation on strings; it
+need not try to determine the existence of a file or directory, and it
+<rfc2119>should not</rfc2119> follow symbolic links. However, two
+pieces of information need to be known from the environment: Whether
+the operating system identifies as “Windows” and the value of the file
+separator. More precisely, the operating system identifies as Windows
+if the <literal>os-name</literal> property as returned by the
+<tag>p:os-info</tag> steps starts with the string
+“<literal>Windows</literal>”. The file separator is what
+<tag>p:os-info</tag> returns as the <literal>file-separator</literal>
+property. If either of them are not known, it is assumed that the
+operating system is not Windows and the file separator is the forward
+slash, “<literal>/</literal>”.</para>
+
+<para>The comparisons and regular expressions that follow are
+presented in lower case, but all of the operations performed are case
+blind: “file”, “FILE”, “File”, and “FiLe” are all identical.
+If the system file separator is not “/”, then all occurrences of the
+file separator are replaced by “/” before beginning the analysis.
+</para>
+
+<section xml:id="urify-analysis">
+<title>Analysis</title>
+
+<para>The <code>filepath</code> presented is analyzed to identify the
+following features:</para>
 
-      <para>The <function>p:urify</function> function behaves normally during static
-        analysis.</para>
-    </section>
+<itemizedlist>
+<listitem>
+<para>The <emphasis>scheme</emphasis>, which may be absent or
+implicitly known or explicitly known.</para>
+</listitem>
+<listitem>
+<para>Whether or not the string can be interpreted as
+<emphasis>hierarchical</emphasis>.</para>
+</listitem>
+<listitem>
+<para>The <emphasis>authority</emphasis>, which may be absent.</para>
+</listitem>
+<listitem>
+<para>The <emphasis>drive letter</emphasis>, which may be absent.</para>
+</listitem>
+<listitem>
+<para>And the <emphasis>path</emphasis>, which may be <emphasis>absolute</emphasis>
+or <emphasis>relative</emphasis>.</para>
+</listitem>
+</itemizedlist>
+
+<para>The analysis proceeds along the following lines, stopping as
+soon as the features have been identified.</para>
+
+<orderedlist>
+<listitem>
+<para>If the <code>filepath</code> is the empty string, it has no
+scheme, no authority, and no drive letter. Its path is the empty
+string and it is relative and hierarchical.</para>
+</listitem>
+<listitem>
+<para>If the <code>filepath</code> is the string “//”, it has no
+scheme, no authority, and no drive letter. Its path is the
+string “/” and it is absolute and hierarchical.</para>
+</listitem>
+<listitem>
+<para>On a Windows system, if the <code>filepath</code> matches the
+regular expression “<code>^(file:/*)?([a-z]):(.*)</code>”, it is a
+“file” scheme URI. If the first match group begins “file:”, it is an
+explicit file scheme URI, otherwise it is an implicit file scheme URI.
```

Comment:
Again I think that it hinges on the normalized `basedir` value. If its scheme is “ftp”, then the “u” in a `filepath` of “u:pass@ftp.acme.com/dir” shouldn’t be interpreted as a drive letter. 

Of course we can define `p:urify()` otherwise, but I think it will be more generic and consistent if file heuristics only apply to `filepath` if `basedir` is known to explicitly or implicitly represent a file URI.

Reply:
I think I can make the change you are suggesting, but I'm not sure it's worth the added complexity. As specified now,
```
p:urify("u:pass@ftp.acme.com/dir", "ftp://ftp.example.com/")
```
is an error (`err:XD0075` on Windows). If `u:` was not interpreted as a drive letter on Windows in this case, it would *still* be an error, `err:XD0077` just as it is on a non-Windows platform now.

Even if we imagine a yet more complex set of heuristics where the `u:` is not interpreted as a drive letter and not interpreted as a scheme and gets resolved against the base dir, it still wouldn't produce a result that's likely to be useful. The resolved URI would be:
```
ftp://ftp.example.com/u:pass@ftp.acme.come/dir
```
I can just about imagine a set of rules that would produce `ftp://u:pass@ftp.acme.com/dir` but only by adding a whole raft of special cases for different patterns and different schemes. That way lies madness, surely?

Is there a case where the current spec produces an error, the change suggested would remove the error without introducing a new one, **and** the result would be what the user was hoping for?